fuse_core/src/callback_wrapper.cpp

```diff
@@ -0,0 +1,148 @@
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <fuse_core/callback_wrapper.h>
+
+namespace fuse_core
+{
+
+
+  CallbackAdapter::CallbackAdapter(std::shared_ptr<rclcpp::Context> context_ptr){
+
+    rcl_guard_condition_options_t guard_condition_options =
+        rcl_guard_condition_get_default_options();
+
+    // Guard condition is used by the wait set to handle execute-or-not logic
+    gc_ = rcl_get_zero_initialized_guard_condition();
+    if (RCL_RET_OK != rcl_guard_condition_init(
+      &gc_, context_ptr->get_rcl_context().get(), guard_condition_options)
+    ) {
+      throw std::runtime_error("Could not init guard condition for callback waitable.");
+    }
+  }
+
+  /**
+   * @brief tell the CallbackGroup how many guard conditions are ready in this waitable
+   */
+  size_t CallbackAdapter::get_number_of_ready_guard_conditions() { return 1;}
+
+
+  /**
+   * @brief tell the CallbackGroup that this waitable is ready to execute anything
+   */
+  bool CallbackAdapter::is_ready(rcl_wait_set_t * wait_set) {
+    (void) wait_set;
+
+    // Prevent multiple simultaneous calls of is_ready (causing the executor to wake multiple times)
+    std::lock_guard<std::mutex> lock(ready_mutex_);
+    return !callback_queue_.empty();
+  }
+
+
+  /**
+   * @brief add_to_wait_set is called by rclcpp during NodeWaitables::add_waitable() and CallbackGroup::add_waitable()
+    waitable_ptr = std::make_shared<CallbackAdapter>();
+    node->get_node_waitables_interface()->add_waitable(waitable_ptr, (rclcpp::CallbackGroup::SharedPtr) nullptr);
+   */
+  void CallbackAdapter::add_to_wait_set(rcl_wait_set_t * wait_set)
+  {
+    std::lock_guard<std::recursive_mutex> lock(reentrant_mutex_);
+    RCLCPP_WARN(rclcpp::get_logger("fuse"),
+    "Could not add callback waitable to wait set.");
+    if (RCL_RET_OK != rcl_wait_set_add_guard_condition(wait_set, &gc_, NULL)) {
+    }
+  }
+
+  /**
+   * @brief check the callback queue and return the next callback to run
+   *
+   */
+  std::shared_ptr< void > CallbackAdapter::take_data(){
+    std::shared_ptr<CallbackWrapperBase> cb_wrapper = nullptr;
+    // fetch the callback ptr and release the lock without spending time in the callback
+    {
+      std::lock_guard<std::recursive_mutex> lock(queue_mutex_);
```

Comment:
It looks like `std::mutex` would do.

Reply:
`std::mutex` is indeed enough, reentrant/recursive is a hangover from when I expected a need for concurrent reads from a multi-threaded executor

Reply:
https://github.com/locusrobotics/fuse/pull/284/commits/cdeca18dee79f9a50a3788c844308a84f192b717